src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = head;
+        } else {
+            tail.next = newNode;
+            newNode.previous = tail;
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexIsValid(index, size, "Trying to add at nonexistent or negative index");
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addAtBeginning(value);
+            return;
+        }
+        addInMiddle(value, index);
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        list.forEach(this::add);
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to get the value at non "
+                + "existing or negative index");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to set at nonexistent or negative index");
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        T oldValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to remove at nonexistent or negative index");
+        if (index == 0) {
+            return removeFirstElement();
+        }
+        if (index == size - 1) {
+            return removeLastElement();
+        }
+        return removeFromMiddle(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        if (head.value == object || head.value.equals(object)) {
+            removeFirstElement();
+            return true;
+        }
+
+        while (temp != null) {
+            if (temp.value == null || temp.value.equals(object)) {
```

Comment:
same here

Reply:
Here we will not get NullPointer as we check temp.value == null first